### DISCUSSION TITLE
TypedColumnView with button on column, how to connect signal?

### DISCUSSION THREAD

**Discussion by miguelbaldi** [Relm4/Relm4#631] [Q&A] [answered]
I'm trying to use a `gtk::CheckButton` on a `RelmColumn`  and I followed the example:  https://github.com/Relm4/Relm4/blob/main/examples/typed_column_view.rs .

But so far I could not see how to connect a signal to the widgets inside a `RelmColumn` without access to some form of `Sender`.
All the lists with widgets examples show dummy widgets without actions (signals).

Is it possible?
Thanks

**Comment by AaronErhardt** [comment, marked as answer]
Yes, you can either clone the sender of a component and give each element a clone of the sender or you can use something like a [MessageBroker](https://docs.rs/relm4/latest/relm4/component/struct.MessageBroker.html#method.send) to send your messages to.

**Reply by miguelbaldi** [reply]
Thanks, Aaron!
I followed your suggestion, and worked like a charm.
I've cloned the component sender and passed to my ListItem's constructor. I've needed to keep a reference to the signal handler id.
The signal setup, I've made using bind/unbind functions of the RelmColumn.
Pseudo-code:
```rust
pub struct MyListItem {
    value: bool,
    sender: ComponentSender<MyComponentModelMsg>, // cloned component sender
    clicked_handler_id: RefCell<Option<SignalHandlerId>>, // reference to the signal, returned by connect_* function
}
pub struct MyColumn;

impl RelmColumn for MyColumn {
    type Root = gtk::CheckButton;
    type Widgets = ();
    type Item = MyListItem;

    const COLUMN_NAME: &'static str = "MyColumn";
    const ENABLE_RESIZE: bool = false;
    const ENABLE_EXPAND: bool = false;

    fn setup(_item: &gtk::ListItem) -> (Self::Root, Self::Widgets) {
        let button = gtk::CheckButton::builder().build();
        (button, ())
    }

    fn bind(item: &mut Self::Item, _: &mut Self::Widgets, button: &mut Self::Root) {
        button.set_active(item.value);
        let sender = item.sender.clone();
        let signal_id = button.connect_toggled(move |b| {
            sender.input(MyComponentModelMsg::ValueToggled(b.is_active()));
        });
        item.clicked_handler_id = RefCell::new(Some(signal_id));
    }
    fn unbind(item: &mut Self::Item, _: &mut Self::Widgets, button: &mut Self::Root) {
        if let Some(id) = item.clicked_handler_id.take() {
            button.disconnect(id);
        };
    }
}
```
